### PR TITLE
[aot_autograd] proper handling for when outputs are aliased but have identical size/stride/offset metadata

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/100348, see the discussion in the issue for details. The problem was that for code like this:
```
def f(x):
    out = ...
    return out, out.detach()
```

The `.detach()` would turn into a `.alias()`, and inductor turns `.alias()` calls into no-ops. Inductor would effectively see that the two graph outputs have the same metadata, and return `out, out`. cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire @ngimel alternatively we could have inductor try to detect when it's not ok to make `.alias()` a no-op, but that would probably require some custom logic instead of making `.alias()` a decomposition.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100115
* #100431
* __->__ #100430

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -591,6 +591,9 @@ def has_same_metadata(t1, t2):
         t1.size() == t2.size()
         and t1.stride() == t2.stride()
         and t1.storage_offset() == t2.storage_offset()
+        and t1.storage_offset() == t2.storage_offset()
```
Comment (ngimel) [inline torch/_functorch/aot_autograd.py:594]: duplicate line?

**Review by ngimel — APPROVED** [review]

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -843,6 +853,20 @@ def inner(*flat_args):
...
                             intermediate_bases.append(o._base)
+            elif (
+                # See https://github.com/pytorch/pytorch/issues/100348 for this case.
+                # This protects against the specific case where a user fn returns (output, output.detach())
+                out_tensor_alias_counts[curr_storage] > 1
+                and len(outs_with_identical_metadata_that_require_grad) > 0
+                and not o.requires_grad
+            ):
+                assert len(outs_with_identical_metadata_that_require_grad) > 0
+                # In theory we could use any of these tensors to regenerat the aliased outputs from,
```
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:864]: regenerate
